idaes/core/solvers/tests/test_petsc.py

```diff
@@ -988,3 +988,161 @@ def diff_eq_rule(m, t):
+    )
+    # No value assigned at 0 because there's no corresponding discretization equation
+    assert m.dxdt[0].value is None
+    assert pyo.value(m.dxdt[1]) == pytest.approx(-0.7580125427537873, rel=1e-3)
+    assert m.dxdt[2].value is None
+
+
+@pytest.mark.unit
+@pytest.mark.skipif(not petsc.petsc_available(), reason="PETSc solver not available")
+def test_calculate_derivatives_integrate_second_half():
+    m = pyo.ConcreteModel()
+
+    m.time = pyodae.ContinuousSet(initialize=(0.0, 1.0, 2.0))
+    m.x = pyo.Var(m.time)
+    m.u = pyo.Var(m.time)
+    m.dxdt = pyodae.DerivativeVar(m.x, wrt=m.time)
+
+    def diff_eq_rule(m, t):
+        return m.dxdt[t] == m.x[t] ** 2 - m.u[t]
+
+    m.diff_eq = pyo.Constraint(m.time, rule=diff_eq_rule)
+
+    discretizer = pyo.TransformationFactory("dae.finite_difference")
+    discretizer.apply_to(m, nfe=1, scheme="BACKWARD")
+
+    m.u.fix(1.0)
+    m.x[1].fix(-0.7580125427537873)
+    m.diff_eq[0].deactivate()
+
+    res = petsc.petsc_dae_by_time_element(
+        m,
+        time=m.time,
+        between=[1.0, 2.0],
+        ts_options={
+            "--ts_type": "beuler",
+            "--ts_dt": 3e-2,
+        },
+        skip_initial=True,  # With u and x fixed, no variables to solve for at t0
+        calculate_derivatives=True,
+    )
+    # No value assigned at 0 because there's no corresponding discretization equation
+    assert m.dxdt[0].value is None
+    assert m.dxdt[1].value is None
+    assert pyo.value(m.dxdt[2]) == pytest.approx(-0.2034733131369807, rel=1e-3)
+
+
+@pytest.mark.unit
+@pytest.mark.skipif(not petsc.petsc_available(), reason="PETSc solver not available")
+def test_calculate_derivatives_forward_difference():
+    m = pyo.ConcreteModel()
+
+    m.time = pyodae.ContinuousSet(initialize=(0.0, 1.0, 2.0))
+    m.x = pyo.Var(m.time)
+    m.u = pyo.Var(m.time)
+    m.dxdt = pyodae.DerivativeVar(m.x, wrt=m.time)
+
+    def diff_eq_rule(m, t):
+        return m.dxdt[t] == m.x[t] ** 2 - m.u[t]
+
+    m.diff_eq = pyo.Constraint(m.time, rule=diff_eq_rule)
+
+    discretizer = pyo.TransformationFactory("dae.finite_difference")
+    discretizer.apply_to(m, nfe=1, scheme="FORWARD")
+
+    m.u.fix(1.0)
+    m.x[0].fix(0.0)
+
+    res = petsc.petsc_dae_by_time_element(
+        m,
+        time=m.time,
+        between=[0.0, 2.0],
+        ts_options={
+            "--ts_type": "beuler",
+            "--ts_dt": 3e-2,
+        },
+        skip_initial=True,  # With u and x fixed, no variables to solve for at t0
+        calculate_derivatives=True,
+    )
+    # It should backfill values for interpolated points like t=1
+    assert pyo.value(m.dxdt[0]) == pytest.approx(-0.7580125427537873, rel=1e-3)
+    assert pyo.value(m.dxdt[1]) == pytest.approx(-0.2034733131369807, rel=1e-3)
+    # No value assigned at 2 because there's no corresponding discretization equation
+    # TODO I don't understand how a value is getting assigned here
+    # assert m.dxdt[2].value is None
```

Comment:
@eslickj , could you help me understand why we're getting derivative values populated here? It's definitely not through the `calculate_derivatives` function, but something being returned by the PETSc solver interface.

Reply:
I think the problem is that with the forward finite difference discretization you should technically be deactivating `m.diffeq[2]` however this runs afoul of the `representative_time` infrastructure in the petsc solver. Seems like there is an assumption baked into the petsc solver logic that the model will be discretized using an implicit approach. Which makes sense since our petsc docs explicitly mention that only "implicit TS solvers are supported".

Reply:
I've tried deactivating `m.diffeq[2]`, and it still gets populated. It isn't populated by the `calculate_derivatives` function, but directly from the solver.